### PR TITLE
storage/mvcc/txn: make tiny refactor for `prewrite`

### PR DESCRIPTION
**Pull request by AndreMouche**
Signed-off-by: shirly <AndreMouche@126.com>

## What have you changed? (mandatory)
This PR make tiny refactor for the operate `prewrite`, in order to make the related part more friendly to the reviewers.

## What are the type of the changes? (mandatory)

Engineering

## How has this PR been tested? (mandatory)

unit test

## Does this PR affect documentation (docs) or release note? (mandatory)

no

## Does this PR affect tidb-ansible update? (mandatory)

no

### CODE REVIEW

File: src/storage/mvcc/txn.rs [hunk 1]
```diff
@@ -161,20 +161,22 @@ impl<S: Snapshot> MvccTxn<S> {
...
-                || (write.write_type != WriteType::Delete
-                    && self.key_exist(&key, write.start_ts - 1)?)
-            {
-                return Err(Error::AlreadyExist { key: key.to_raw()? });
-            }
+        if !should_not_exist || write.write_type == WriteType::Delete {
+            return Ok(());
+        }
+
+        if write.write_type == WriteType::Put || self.key_exist(&key, write.start_ts - 1)? {
```
Comment (MyonKeminta) [inline src/storage/mvcc/txn.rs]: How about adding a comment to describe why here it calls `key_exist`?

File: src/storage/mvcc/txn.rs [hunk 2]
```diff
@@ -363,58 +359,54 @@ impl<S: Snapshot> MvccTxn<S> {
...
-                        primary: lock.primary,
-                        ts: lock.ts,
-                        ttl: lock.ttl,
-                        txn_size: lock.txn_size,
-                    });
-                }
-                // TODO: remove it in future
-                if lock.lock_type == LockType::Pessimistic {
-                    return Err(Error::LockTypeNotMatch {
+        // for the insert operation, the old key should not in the system.
```
Comment (MyonKeminta) [inline src/storage/mvcc/txn.rs]: Consider capitalize the first letter of these comments

Comment (MyonKeminta) [reply src/storage/mvcc/txn.rs]:
```suggestion
        // For the insert operation, the old key should not be in the system.
```

**Comment by breeswish** [comment]
Note: The type of this PR is "Engineering".

File: src/storage/mvcc/txn.rs [hunk 3]
```diff
@@ -363,58 +359,54 @@ impl<S: Snapshot> MvccTxn<S> {
...
+        let should_not_exist = mutation.is_insert();
+        let (key, value) = mutation.into_key_value();
+        // check whether there is a newer version.
```
Comment (youjiali1995) [inline src/storage/mvcc/txn.rs]: ditto

File: src/storage/mvcc/txn.rs [hunk 4]
```diff
@@ -363,58 +359,54 @@ impl<S: Snapshot> MvccTxn<S> {
...
                     });
                 }
-                // Duplicated command. No need to overwrite the lock and data.
-                MVCC_DUPLICATE_CMD_COUNTER_VEC.prewrite.inc();
-                return Ok(());
+                self.check_data_constraint(should_not_exist, &write, &key)?;
             }
         }
 
+        // check whether the current key is locked at any timestamp.
```
Comment (youjiali1995) [inline src/storage/mvcc/txn.rs]: ditto

File: src/storage/mvcc/txn.rs [hunk 5]
```diff
@@ -161,20 +161,22 @@ impl<S: Snapshot> MvccTxn<S> {
         }
     }
 
+    /// check whether the existence of the key is according to `should_not_exist`.
+    /// If not,returns an AlreadyExist error.
```
[the remark this answers is not in the document]
Comment (youjiali1995) [reply src/storage/mvcc/txn.rs]: Where is the specification of documents?

File: src/storage/mvcc/txn.rs [hunk 6]
```diff
@@ -161,8 +161,8 @@ impl<S: Snapshot> MvccTxn<S> {
         }
     }
 
-    /// check whether the existence of the key is according to `should_not_exist`.
-    /// If not,returns an AlreadyExist error.
+    /// Checks whether the existence of the key is according to `should_not_exist`.
+    /// If not,returns an `AlreadyExist` error.
```
Comment (MyonKeminta) [inline src/storage/mvcc/txn.rs]: One space after English punctuation
```
    /// If not, returns ...
```

File: src/storage/mod.rs [hunk 7]
```diff
@@ -76,6 +76,22 @@ impl Mutation {
...
+        match self {
+            Mutation::Put((key, value)) => (key, Some(value)),
+            Mutation::Delete(key) => (key, None),
+            Mutation::Lock(key) => (key, None),
+            Mutation::Insert((key, value)) => (key, Some(value)),
+        }
+    }
+
+    pub fn is_insert(&self) -> bool {
+        match *self {
```
Comment (lonng) [inline src/storage/mod.rs]:
```suggestion
        match self {
```

File: src/storage/mvcc/txn.rs [hunk 8]
```diff
@@ -161,20 +161,25 @@ impl<S: Snapshot> MvccTxn<S> {
         }
     }
 
+    /// Checks whether the existence of the key is according to `should_not_exist`.
```
Comment (sticnarf) [inline src/storage/mvcc/txn.rs]: It now looks weird. Maybe `Checks whether the key exists according...` or `Checks the existence of the key according...` would be better

File: src/storage/mvcc/txn.rs [hunk 9]
```diff
@@ -161,20 +161,25 @@ impl<S: Snapshot> MvccTxn<S> {
...
-                || (write.write_type != WriteType::Delete
-                    && self.key_exist(&key, write.start_ts - 1)?)
-            {
-                return Err(Error::AlreadyExist { key: key.to_raw()? });
-            }
+        if !should_not_exist || write.write_type == WriteType::Delete {
+            return Ok(());
+        }
+
+        // The current key is exist when meet any of the following conditions:
```
Comment (sticnarf) [inline src/storage/mvcc/txn.rs]:
```suggestion
        // The current key exists under any of the following conditions:
```

File: src/storage/mod.rs [hunk 10]
```diff
@@ -69,13 +69,29 @@ pub enum Mutation {
 #[allow(clippy::match_same_arms)]
 impl Mutation {
     pub fn key(&self) -> &Key {
-        match *self {
+        match self {
```
Comment (zhouqiang-cl) [inline src/storage/mod.rs]: why remove *

Comment (AndreMouche) [reply src/storage/mod.rs]: Because we do not need it.

Comment (sticnarf) [reply src/storage/mod.rs]: It's just unnecessary after an ergonomics improvement.
https://github.com/rust-lang/rfcs/blob/master/text/2005-match-ergonomics.md

**Pull request merged**